### PR TITLE
Upgrade rubocop to 0.49.1

### PR DESCRIPTION
**Pull request by mattnichols**
This brings the ruby parser up-to-date. New cops `Rails/HttpPositionalArguments` and `Style/FrozenStringLiteralComment` are disabled and some cop names were fixed.

@ztoolson @mmmries @liveh2o @brianstien 

Here are some samples of offense reports to give an idea of how much work we will need to do to upgrade:

__Note: I left `Rails/HttpPositionalArguments` enabled for these reports to show how wide-spread the fix would be. I think that enabling `Style/FrozenStringLiteralComment` would be okay since there is an auto-fix for it and it is only a ruby 3.0 compatibility comment.__

```
331  Rails/HttpPositionalArguments
272  Metrics/BlockLength
141  Style/StringLiterals
28   Layout/AlignParameters
22   Rails/Blank
12   Rails/ApplicationRecord
6    Style/EmptyMethod
4    Style/PercentLiteralDelimiters
3    Bundler/OrderedGems
3    Rails/SkipsModelValidations
2    Layout/IndentHeredoc
2    Layout/TrailingBlankLines
2    Security/JSONLoad
2    Style/EmptyCaseCondition
1    Layout/SpaceBeforeBlockBraces
1    Lint/ScriptPermission
1    Rails/ApplicationJob
1    Rails/FilePath
1    Rails/Present
1    Style/CommentAnnotation
1    Style/MethodMissing
--
837  Total
```

```
244  Rails/HttpPositionalArguments
42   Rails/DynamicFindBy
14   Rails/Blank
12   Rails/ApplicationRecord
12   Style/PercentLiteralDelimiters
10   Layout/SpaceBeforeBlockBraces
6    Style/EmptyMethod
6    Style/StringLiterals
4    Rails/FilePath
3    Style/ClassAndModuleChildren
3    Style/NumericPredicate
2    Layout/SpaceInsidePercentLiteralDelimiters
2    Rails/OutputSafety
2    Security/YAMLLoad
1    Bundler/OrderedGems
1    Layout/EmptyLinesAroundExceptionHandlingKeywords
1    Layout/IndentHeredoc
1    Layout/MultilineMethodCallBraceLayout
1    Layout/SpaceBeforeFirstArg
--
367  Total
```

```
694   Rails/HttpPositionalArguments
403   Style/StringLiterals
259   Metrics/BlockLength
66    Rails/Blank
58    Style/EmptyMethod
43    Style/PercentLiteralDelimiters
32    Rails/OutputSafety
21    Lint/AmbiguousBlockAssociation
12    Rails/ApplicationRecord
11    Layout/MultilineMethodCallBraceLayout
11    Rails/SkipsModelValidations
6     Style/VariableNumber
5     Layout/MultilineHashBraceLayout
5     Lint/UnifiedInteger
5     Rails/Present
4     Bundler/OrderedGems
4     Style/InverseMethods
4     Style/NumericPredicate
3     Layout/SpaceBeforeBlockBraces
3     Style/EmptyCaseCondition
3     Style/GuardClause
2     Layout/EmptyLinesAroundExceptionHandlingKeywords
2     Layout/SpaceAroundOperators
2     Layout/SpaceInsidePercentLiteralDelimiters
2     Style/MultilineIfModifier
1     Layout/IndentHeredoc
1     Layout/MultilineArrayBraceLayout
1     Lint/ScriptPermission
1     Lint/UnneededDisable
1     Performance/CompareWithBlock
1     Rails/DynamicFindBy
1     Rails/FilePath
1     Style/MultipleComparison
1     Style/RedundantReturn
--
1669  Total
```

### CODE REVIEW

File: lib/disabled_cops.yml [hunk 1]
```diff
@@ -48,18 +61,10 @@ Style/EachWithObject:
...
-Style/FirstHashElementLineBreak:
-  Enabled: false
-Style/FirstMethodArgumentLineBreak:
-  Enabled: false
-Style/FirstMethodParameterLineBreak:
-  Enabled: false
 Style/FormatString:
   Enabled: false
+Style/FrozenStringLiteralComment:
+  Enabled: false
```
Comment (ryanbjones) [inline lib/disabled_cops.yml:77]: 👍  from me on this one being false...I really don't care for it being at the top of every class even if it is auto-correctable

**Comment by ztoolson** [comment]
@mattnichols Previously I have disabled all Metric specs by default, and if we wanted to worry about that we could opt in (with reference to `Metrics/BlockLength`).

**Comment by ryanbjones** [comment]
The `Metrics/BlockLength` I think should be disabled because it really is somewhat arbitrary and I think can make code more confusing with the way we sometimes try to get around it

**Comment by brianstien** [comment]
@mattnichols According to documentation [HttpPositionalArguments is only for rails >= 5](http://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Rails/HttpPositionalArguments). We should opt-in on rails 5 projects, but default should be off to make life easier for rails 4 projects.

edit:
Alternatively they say you can set "TargetRailsVersion"... I am not familiar but maybe we can try it?

**Comment by mattnichols** [comment]
I was just about to look into the Metrics/BlockLength cop. It isn't new (as far as I can tell), but seems to be counting different now. I'm good with disabling for specs. I didn't change any of that... hmmm... 👀

**Comment by ryanbjones** [comment]
I feel like we are pretty deliberate about when we skip model validations/callbacks so I'd be fine to disable it. Perhaps it's only use would be to force us to comment that it was occurring so you could easily search for all instances where we skip callbacks

**Comment by ryanbjones** [comment]
The `Rails/Blank` cop would change quite a few spots where we use `unless resource.present?` since it prefers to use `empty?` or `blank?` to avoid using `unless`..I'm pretty used to our `unless` format and would be fine to keep our usage that way
http://rubocop.readthedocs.io/en/latest/cops_rails/#railsblank

File: lib/modified_cops.yml [hunk 2]
```diff
@@ -20,3 +20,7 @@ Style/StringLiterals:
   Description: Checks if uses of quotes match the configured preference.
   Enabled: true
   EnforcedStyle: double_quotes
+
+Metrics/BlockLength:
```
Comment (ztoolson) [inline lib/modified_cops.yml]: We can't exclude folders from mad rubocop. The gem can't assume the folder structure of the project. I think this cop should just be turned off.

Comment (mattnichols) [reply lib/modified_cops.yml]: @ztoolson Done

**Comment by mattnichols** [comment]
## Updated stats

These are the offenses from the same projects after suggested updates

```
141  Style/StringLiterals
28   Layout/AlignParameters
25   Rails/Blank
8    Style/EmptyMethod
4    Style/PercentLiteralDelimiters
3    Rails/SkipsModelValidations
2    Bundler/OrderedGems
2    Layout/IndentHeredoc
2    Layout/TrailingBlankLines
2    Security/JSONLoad
2    Style/EmptyCaseCondition
1    Layout/SpaceBeforeBlockBraces
1    Lint/ScriptPermission
1    Rails/FilePath
1    Rails/Present
1    Style/CommentAnnotation
1    Style/MethodMissing
--
225  Total
```

```
42   Rails/DynamicFindBy
16   Rails/Blank
11   Style/PercentLiteralDelimiters
10   Layout/SpaceBeforeBlockBraces
6    Style/EmptyMethod
6    Style/StringLiterals
4    Rails/FilePath
3    Style/ClassAndModuleChildren
3    Style/NumericPredicate
2    Layout/SpaceInsidePercentLiteralDelimiters
2    Rails/OutputSafety
2    Security/YAMLLoad
1    Bundler/OrderedGems
1    Layout/EmptyLinesAroundExceptionHandlingKeywords
1    Layout/IndentHeredoc
1    Layout/MultilineMethodCallBraceLayout
1    Lint/AmbiguousBlockAssociation
--
112  Total
```

```
405  Style/StringLiterals
66   Rails/Blank
58   Style/EmptyMethod
43   Style/PercentLiteralDelimiters
32   Rails/OutputSafety
21   Lint/AmbiguousBlockAssociation
11   Layout/MultilineMethodCallBraceLayout
11   Rails/SkipsModelValidations
6    Style/VariableNumber
5    Layout/MultilineHashBraceLayout
5    Lint/UnifiedInteger
5    Rails/Present
4    Bundler/OrderedGems
4    Style/InverseMethods
4    Style/NumericPredicate
3    Layout/SpaceBeforeBlockBraces
3    Style/EmptyCaseCondition
3    Style/GuardClause
2    Layout/EmptyLinesAroundExceptionHandlingKeywords
2    Layout/SpaceAroundOperators
2    Layout/SpaceInsidePercentLiteralDelimiters
2    Style/MultilineIfModifier
1    Layout/IndentHeredoc
1    Layout/MultilineArrayBraceLayout
1    Lint/ScriptPermission
1    Lint/UnneededDisable
1    Performance/CompareWithBlock
1    Rails/DynamicFindBy
1    Rails/FilePath
1    Style/MultipleComparison
1    Style/RedundantReturn
--
706  Total
```

**Comment by mattnichols** [comment]
I propose removal of https://github.com/bbatsov/rubocop/blob/a61bcd3a22d471c7947ae3c6c0045fc4854ebbe3/manual/cops_style.md#stylenumericpredicate

Here is an example where I think it can make the code less readable: 

```
app/models/status.rb:7:44: C: Style/NumericPredicate: Use Status.count.positive? instead of Status.count > 0.
    fail "There can only be one Status" if Status.count > 0
```

This cop will enforce all predicates or all comparisons. I would prefer to have a choice based on the situation.

@ztoolson @brianstien @ryanbjones

**Comment by ryanbjones** [comment]
👍  from me at this point

**Review by ztoolson — APPROVED** [review]

**Pull request merged**